### DISCUSSION TITLE
MasterDetailRelation into MasterDetailRelation

### DISCUSSION THREAD

**Discussion by pesnunj0** [serenity-is/Serenity#5959] [Help] [answered]
Hi,
I have a MasterDetail file into another MasterDetail. So I have a cascaded MasterDetail 
So I have Company with id=CompanyId, Incidents with Id =IncidentIId  and foreignkey=CompanyId and Notifications with Id=NotificactionId and foreignkey=IncidentId.
First level MasterDetail works fine but second level (Notificactions) doesn´t
The behavior is as follows:
 1.- Columns for Notifications are not displayed. (I use IncludeColumns
 2.- When I add a new Notification within and incident it is added into table but the rest of notifications with the same IncidentId are deleted

I followed instructions detailed in tutorial "The Cast and Characters They Played" for both Master Detail
I probably made a mistake but I can't find it.

Thanks in advance

-

**Comment by edwardch** [comment, marked as answer]
I've had to do something like this, and it took me a while to make sure everything was included. 

Check the levels, something like so

Top Level

CompanyRow.cs

```cs
[DisplayName("Incidents"), MasterDetailRelation(foreignKey: "CompanyId", IncludeColumns = "NotificationList"), NotMapped]
public List<IncidentRow> IncidentList
{
    get { return Fields.IncidentList[this]; }
    set { Fields.IncidentList[this] = value; }
}
```

CompanyForm.cs
```cs
[IncidentEditor]
public List<Entities.IncidentRow> IncidentList { get; set; }
```

Second Level
IncidentRow.cs

```cs
// Assuming a notification has Timestamp and Text columns that you want displayed in the grid
[DisplayName("Notifications"), MasterDetailRelation(foreignKey: "IncidentIId", IncludeColumns = "TimeStamp, Text"), NotMapped]
public List<NotificationRow> NotificationList
{
    get { return Fields.NotificationList[this]; }
    set { Fields.NotificationList[this] = value; }
}
```

IncidentForm.cs
```cs
[NotificationEditor]
public List<Entities.NotificationRow> NotificationList { get; set; }
```

IncidentEditor.ts
```ts
/// <reference path="../../Common/Helpers/GridEditorBase.ts" />

namespace YourProject.YourModule {

    @Serenity.Decorators.registerClass()
    export class IncidentEditor extends Common.GridEditorBase<IncidentRow> {
        protected getColumnsKey() { return "YourModule.Incident"; }
        protected getDialogType() { return IncidentDialog; }
        protected getLocalTextPrefix() { return IncidentRow.localTextPrefix; }

        constructor(container: JQuery) {
            super(container);
        }

        protected getAddButtonCaption() {
            return "Add";
        }

        validateEntity(row, id) {
            // row.Type = Q.toId(row.Type);

            return true;
        }
    }
}
```


Third/Lowest Level
NotificationEditor.ts

```ts
/// <reference path="../../Common/Helpers/GridEditorBase.ts" />

namespace YourProject.YourModule {

    @Serenity.Decorators.registerClass()
    export class NotificationEditor extends Common.GridEditorBase<NotificationRow> {
        protected getColumnsKey() { return "YourModule.Notification"; }
        protected getDialogType() { return NotificationDialog; }
        protected getLocalTextPrefix() { return NotificationRow.localTextPrefix; }

        constructor(container: JQuery) {
            super(container);
        }

        protected getAddButtonCaption() {
            return "Add";
        }

        validateEntity(row, id) {
            return true;
        }
    }
}
```

Create the TypeScript editors first, run the project, and only then add the [xEditor] annotations to your form fields.

HTH!

**Reply by wirble** [reply]
Thanks for posting this.  Was struggling with same issue.  Missing the IncludeColumns "NotificationList" on the parent.  Who would have thought to put the child's child (Grandchild) list into the IncludeColumns in the first level.  I guess I don't fully understand the purpose of IncludeColumns...

**Comment by pesnunj0** [comment]
Hi Edward,
Thank you so much for your kind and quick answer.
My code is I think quite the same.
Assuming firsdt level is ok becouse it works. I write you second and third level:

Second level Form:
       [EnviosEditor]
        public List<Entities.EnviosRow> EnviosList { get; set; }

Second level Row:
        [MasterDetailRelation(foreignKey: "HistorialId", IncludeColumns = "TipoEnvioTipo,EstadoEnvioEstado,FormaEnvioForma,ContactoEnvioFullName,FechaEnvio"), NotMapped]

        [DisplayName("Envios")]
        public List<EnviosRow> EnviosList
        {
            get { return Fields.EnviosList[this]; }
            set { Fields.EnviosList[this] = value; }
        }
Third level editor (enviosEdito.ts):
/// <reference path="../../Common/Helpers/GridEditorBase.ts" />


namespace ProyectosZec.Nuevo_Roezec {

    @Serenity.Decorators.registerClass()
    export class EnviosEditor extends Common.GridEditorBase<EnviosRow> {
        /*protected getPersistanceStorage(): Serenity.SettingStorage { return new Common.UserPreferenceStorage(); }*/
        protected getColumnsKey() { return "Nuevo_Roezec.Envios"; }
        protected getDialogType() { return EnviosEditDialog; }
        protected getLocalTextPrefix() { return EnviosRow.localTextPrefix; }

        constructor(container: JQuery) {
            super(container);
        }

        protected getAddButtonCaption() {
            return "Añadir Envío";
        }
    }
}
Now EnviosEditDialog.ts

/// <reference path="../../Common/Helpers/GridEditorDialog.ts" />

namespace ProyectosZec.Nuevo_Roezec {

    @Serenity.Decorators.registerClass()
    export class EnviosEditDialog extends Common.GridEditorDialog<EnviosRow> {
        protected getFormKey() { return EnviosForm.formKey; }
        protected getNameProperty() { return EnviosRow.nameProperty; }
        protected getLocalTextPrefix() { return EnviosRow.localTextPrefix; }

        protected form: EnviosForm;

        constructor() {
            super();
            this.form = new EnviosForm(this.idPrefix);
        }
    }
}

I think it is quite the same as u but .... I' ll try to review, but if you see where my mistake is I would thank so much
Thanks in advance
Javier

**Comment by pesnunj0** [comment]
Hi All,
Solved the issue Adding MinSelectLevel(SelectLevel.List) to NotMapped variable in second level Row file
Thanks
